server/src/lib/model/db.ts

```diff
@@ -194,26 +323,86 @@ export default class DB {
         SELECT clips.*
         FROM clips
         LEFT JOIN sentences on clips.original_sentence_id = sentences.id
-        WHERE is_valid IS NULL AND clips.locale_id = ? AND client_id <> ? AND
-              NOT EXISTS(
-                SELECT *
-                FROM votes
-                WHERE votes.clip_id = clips.id AND client_id = ?
-              )
+        WHERE is_valid IS NULL AND clips.locale_id = ? AND client_id <> ?
+        AND NOT EXISTS(
+            SELECT *
+            FROM votes
+            WHERE votes.clip_id = clips.id AND client_id = ?
+          )
+        AND (sentences.clips_count <= 10 OR sentences.clips_count IS NULL)
         ORDER BY sentences.clips_count ASC, clips.created_at ASC
         LIMIT ?
       ) t
       ORDER BY RAND()
-      LIMIT ?
-    `,
-      [await getLocaleId(locale), client_id, client_id, SHUFFLE_SIZE, count]
+      LIMIT ?`,
+      [locale_id, client_id, client_id, SHUFFLE_SIZE, count]
     );
     for (const clip of clips) {
       clip.voters = clip.voters ? clip.voters.split(',') : [];
     }
     return clips as DBClipWithVoters[];
   }
 
+  async findClipsMatchingTaxonomy(
+    client_id: string,
+    locale_id: number,
+    count: number,
+    term_name: string
+  ): Promise<DBClipWithVoters[]> {
+    const [clips] = await this.mysql.query(
+      `
+      SELECT *
+      FROM (
+        SELECT * FROM clips WHERE original_sentence_id IN (
+          SELECT sentence_id FROM taxonomy_entries
+          LEFT JOIN sentences ON taxonomy_entries.sentence_id = sentences.id
+          WHERE locale_id = ?
+          AND sentence_id NOT IN (
+            SELECT original_sentence_id FROM (
+              SELECT original_sentence_id, clips.is_valid, count(votes.id) as vote_count
+              FROM clips
+              LEFT JOIN votes ON votes.clip_id = clips.id
+              LEFT JOIN taxonomy_entries entries
+                ON clips.original_sentence_id = entries.sentence_id
+              WHERE clips.locale_id = ?
+              AND (votes.client_id = ?)
+              AND entries.term_id = ?
+              GROUP BY original_sentence_id
+              HAVING vote_count >= 2
+            ) vote_counts
+          )
+        )
```

Comment:
This is the biggest change in actual logic compared to the first PR, and it's also because during testing this query ground to a halt if you had already voted hundreds of thousands of times, or had tens of thousands of clips. The logic is kind of backwards - it looks for a list of sentences that match the taxonomy that the user has already voted twice on, and removes it from a full list of sentences matching taxonomy. Because it's primarily selecting for taxonomy_entries though, it's super fast.